internal/vault_reader.go

```diff
@@ -0,0 +1,293 @@
+	}
+
+	return ClusterCredentials{
+		data: map[string][]byte{
+			"username": []byte(username),
+			"password": []byte(password),
+		},
+	}, nil
+}
+
+func getValue(key string, data map[string]interface{}) (string, error) {
+	result, ok := data[key].(string)
+	if !ok {
+		return "", fmt.Errorf("expected %s to be a string but is a %T", key, data[key])
+	}
+
+	return result, nil
+}
+
+func availableKeys(m map[string]interface{}) []string {
+	result := make([]string, len(m))
+	i := 0
+	for k := range m {
+		result[i] = k
+		i++
+	}
+	return result
+}
+
+func login(vaultClient *vault.Client, vaultSpec *rabbitmqv1beta1.VaultSpec) (*vault.Secret, error) {
+	logger := ctrl.LoggerFrom(nil)
+
+	// GCH TODO return to this...
+	// role := vaultSpec.Role
+	// if role == "" {
+	// 	return nil, errors.New("no role value set in Vault secret backend")
+	// }
+	role := "messaging-topology-operator"
+
+	var annotations = vaultSpec.Annotations
+	if annotations["vault.hashicorp.com/namespace"] != "" {
+		vaultClient.SetNamespace(annotations["vault.hashicorp.com/namespace"])
+	}
+
+	jwt, err := ReadServiceAccountTokenFunc()
+	if err != nil {
+		return nil, fmt.Errorf("unable to read file containing service account token: %w", err)
+	}
+
+	loginAuthPath := defaultAuthPath
+	annotations = vaultSpec.Annotations
+	if annotations["vault.hashicorp.com/auth-path"] != "" {
+		loginAuthPath = annotations["vault.hashicorp.com/auth-path"]
+	}
+
+	logger.Info("Authenticating to Vault")
+
+	vaultSecret, err := ReadVaultClientSecretFunc(vaultClient, string(jwt), role, loginAuthPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to obtain Vault client secret: %w", err)
+	}
+
+	if vaultSecret == nil || vaultSecret.Auth == nil || vaultSecret.Auth.ClientToken == "" {
+		return nil, fmt.Errorf("no client token found in Vault secret")
+	}
+
+	vaultClient.SetToken(vaultSecret.Auth.ClientToken)
+	return vaultSecret, nil
+}
+
+func renewToken(client *vault.Client, vaultSpec *rabbitmqv1beta1.VaultSpec, initialLoginErrorCh chan<- error) {
+	logger := ctrl.LoggerFrom(nil)
+	sentFirstLoginAttemptErr := false
+
+	for {
+		vaultLoginResp, err := login(client, vaultSpec)
+		if err != nil {
+			logger.Error(err, "unable to authenticate to Vault server")
+		}
+
+		if !sentFirstLoginAttemptErr {
+			initialLoginErrorCh <- err
+			sentFirstLoginAttemptErr = true
+			if err != nil {
+				// Initial login attempt failed so fail fast and don't try to manage (non-existent) token lifecycle
+				logger.Info("Lifecycle management of Vault token will not be carried out")
+				return
+			}
+			logger.Info("Initiating lifecycle management of Vault token")
+		}
+
+		err = manageTokenLifecycle(client, vaultLoginResp)
+		if err != nil {
+			logger.Error(err, "unable to start managing the Vault token lifecycle")
+		}
+
+		// Reduce load on Vault server in a problem situation where repeated login attempts may be made
+		time.Sleep(2 * time.Second)
```

Comment:
In the future, this could be replaced by an exponential backoff delay. I'd release it just how it is and if necessary change it to exponential backoff delay rather than fixed delay.